irohad/ametsuchi/impl/postgres_block_storage.cpp

```diff
@@ -0,0 +1,118 @@
+       soci::use(block->height()),
+       soci::use(block->blob().hex()));
+  log_->debug("insert: {}", block->blob().hex());
+  try {
+    st.execute(true);
+    return true;
+  } catch (const std::exception &e) {
+    log_->warn(
+        "Failed to insert block {}, reason {}", block->height(), e.what());
+    return false;
+  }
+}
+
+boost::optional<std::shared_ptr<const shared_model::interface::Block>>
+PostgresBlockStorage::fetch(
+    shared_model::interface::types::HeightType height) const {
+  using T = boost::tuple<std::string>;
+  auto result = execute<T>([&] {
+    return (
+        sql_.prepare << "SELECT block_data FROM blocks WHERE height = :height",
+        soci::use(height));
+  });
+  return flatMapValue<
+      boost::optional<std::shared_ptr<const shared_model::interface::Block>>>(
+      result, [&](auto &block_data) {
+        log_->debug("fetched: {}", block_data);
+        auto byte_block = iroha::hexstringToBytestring(block_data);
+        if (not byte_block) {
+          return boost::optional<
+              std::shared_ptr<const shared_model::interface::Block>>(
+              boost::none);
+        }
+
+        iroha::protocol::Block_v1 block;
+        block.ParseFromString(*byte_block);
+        return block_factory_->build(std::move(block))
+            .match(
+                [&](auto &&v) {
+                  return boost::make_optional(
+                      std::shared_ptr<const shared_model::interface::Block>(
+                          std::move(v.value)));
+                },
+                [&](const auto &e)
+                    -> boost::optional<
+                        std::shared_ptr<const shared_model::interface::Block>> {
+                  log_->error("Could not build block at height {}: {}",
+                              height,
+                              e.error.error);
+                  return boost::none;
+                });
+      });
+}
+
+size_t PostgresBlockStorage::size() const {
+  using T = boost::tuple<shared_model::interface::types::HeightType>;
+  auto result = execute<T>(
+      [&] { return (sql_.prepare << "SELECT COUNT(*) FROM blocks"); });
+  return flatMapValue<
+             boost::optional<shared_model::interface::types::HeightType>>(
+             result, [](auto &count) { return boost::make_optional(count); })
+      .value_or(0);
+}
+
+void PostgresBlockStorage::clear() {
+  soci::statement st = sql_.prepare << "TRUNCATE blocks";
+  try {
+    st.execute(true);
+  } catch (const std::exception &e) {
+    log_->warn("Failed to clear blocks table, reason {}", e.what());
+  }
+}
+
+void PostgresBlockStorage::forEach(
+    iroha::ametsuchi::BlockStorage::FunctionType function) const {
+  using T = boost::tuple<shared_model::interface::types::HeightType>;
+  auto result =
+      execute<T>([&] { return (sql_.prepare << "SELECT height FROM blocks"); });
+  auto ids = mapValues<std::vector<shared_model::interface::types::HeightType>>(
```

Comment:
What is the point to create a vector with all possible heights? I think we can fetch only the top height and iterate over `[1..,height] collection, isn't it?

Reply:
Well, it was my initial implementation. The problem is we can't assume block numbers are sequential and start from 1. At least both FlatFile and InMemory implementations don't have this assumption.

Reply:
We have to put range checking on initialization and use it as assumption further.